### DISCUSSION TITLE
A bison grammar for cweb

### DISCUSSION THREAD

**Discussion by zmajeed** [gogotrango/cwebdraftapi#4] [General]
A draft grammar for cweb is at https://github.com/gogotrango/cwebdraftapi/blob/main/grammar/cweb.grammar.bison.y

It is based on the cweb manual and docs for cweb tools - ctangle and cweave - the docs are in https://github.com/gogotrango/cwebdraftapi/tree/main/doc/cwebdoc

This grammar file has just the rules for a bison 3.7 LALR(1) parser - I plan to post a sample of a draft callback API based on this soon.

I'd appreciate any input on how well this grammar represents the cweb language

Tokens in the grammar have string literal names like `"@space untitled tex section"`. The names generally start with the cweb control code followed by its description.

**Comment by texdraft** [comment]
It looks good overall. There are only a few things I noticed:

1. The C part of a section may be introduced by `@p` or `@P`, as well as `@c`/`@c`. But you might be canonicalizing this into `@c`, in which case it doesn't matter.
2. A section may be completely empty. (I wasn't sure about this, so I tested it. The ridiculous input file
    ```
    @ @ @ @ @ @ @ @ @c@
    ```
    is processed without complaint; the `@c` is there only to shut up `ctangle`, which would complain about a lack of C text.)
3. Section names should be allowed in inner C text.
4. A few of the terminal symbols/tokens are not used (such as `@& nbsp`)

**Reply by zmajeed** [reply]
Thanks for looking - you guessed right about `@p` - it's subsumed into`@c` by the lexer - flex in this case

I'll add section names to inner C contexts

I've tried to account for boundary cases especially empty content but I'm sure there's some I'm missing and I appreciate the test samples

**Reply by zmajeed** [reply]
Updated the grammar - added a comment that control codes are case-insensitive - changed the token name to `"@c/@p unnamed program section"` - added rules for the previously unused tokens - allow section names in inner C context

Empty sections were already accepted by the grammar - middle and c sections are covered by the `section` rules

```bison
// section has 1, 2 or 3 parts
// any part can be empty but a nonempty part must appear in correct order
section:
       tex middle c
       | tex middle
       | tex c
       | tex
       ;
```

and the tex section could be empty via the third rule of cweb

```bison
cweb:
    limbo sections
    | sections
    | limbo
    ;
```

each part is allowed empty content, e.g. the tex part can be empty via its second and fourth rules

```bison
tex:
   "@space untitled tex section" tex_contents
   | "@space untitled tex section"
   | "@* titled tex section" tex_contents
   | "@* titled tex section"
   ;
```

but a completely empty file was not recognized before because both limbo and sections could not be empty - now limbo is explicitly allowed to be empty - empty rules are often problematic because they lead to ambiguity but it turns out to be useful here and simplifies the top cweb production

```bison
// cweb document has sections with optional limbo prolog
cweb:
    limbo sections
    | limbo
    ;

// limbo has pieces of limbo content
limbo:
     %empty
     | limbo_contents
     ;
```
